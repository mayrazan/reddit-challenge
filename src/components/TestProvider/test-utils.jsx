import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import theme from '../../theme';
import GlobalStyle from '../../theme/GlobalStyle';
import { lightColors } from '../../theme/colors';

const AllTheProviders = ({ children, ...props }) => (
  <ThemeProvider theme={{ ...theme, ...lightColors }} {...props}>
    <GlobalStyle />
    {children}
  </ThemeProvider>
);
AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired,
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
