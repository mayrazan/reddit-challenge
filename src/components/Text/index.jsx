/* eslint-disable consistent-return */
import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

export const TextStyleVariants = (name) => {
  if (name) {
    return css`
      ${({ theme, bold }) => css`
        font-size: ${theme.typographyFonts[name].fontSize};
        font-weight: ${bold ? 'bold' : theme.typographyFonts[name].fontWeight};
        line-height: ${theme.typographyFonts[name].lineHeight};
      `}
    `;
  }
};

const TextBase = styled.span`
  color: ${({ theme, color }) => color && theme.text[color]};
  margin: 0;
  ${({ variant }) => {
    if (typeof variant === 'string') {
      return TextStyleVariants(variant);
    }
  }}
  ${({ as }) => as === 'a'
    && css`
      cursor: pointer;
      text-decoration: none;
      &:hover {
        color: ${({ theme }) => theme.text.secondary};
      }
    `}
`;

export default function Text({
  variant,
  children,
  tag,
  color,
  bold,
  ...props
}) {
  if (props.href) {
    return (
      <TextBase
        as="a"
        variant={variant}
        href={props.href}
        color={color}
        bold={bold}
        target="_blank"
        {...props}
      >
        {children}
      </TextBase>
    );
  }
  return (
    <TextBase as={tag} variant={variant} color={color} bold={bold} {...props}>
      {children}
    </TextBase>
  );
}

Text.propTypes = {
  tag: PropTypes.string,
  variant: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  children: PropTypes.node.isRequired,
  color: PropTypes.string,
  bold: PropTypes.bool,
  href: PropTypes.string,
};

Text.defaultProps = {
  tag: 'span',
  variant: 'subtitle',
  color: '#000000',
  bold: false,
  href: null,
};
