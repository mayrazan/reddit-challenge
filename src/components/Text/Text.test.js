import React from 'react';
import { render, screen } from '../TestProvider/test-utils';
import Text from '.';

describe('<Text />', () => {
  test('render component with props', () => {
    render(
      <Text tag="a" variant="title" color="contrastText" href="/" bold>
        Home
      </Text>,
    );
    const text = screen.getByRole('link', {
      name: /home/i,
    });
    expect(text).toMatchSnapshot();
  });

  test('render component', () => {
    render(<Text>Home</Text>);

    const text = screen.getByText(/home/i);
    expect(text).toMatchSnapshot();
  });
});
