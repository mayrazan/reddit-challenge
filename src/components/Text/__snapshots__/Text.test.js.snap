// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> render component 1`] = `
.c0 {
  margin: 0;
  font-size: 20px;
  font-weight: 600;
  line-height: 25px;
}

<span
  class="c0"
  color="#000000"
>
  Home
</span>
`;

exports[`<Text /> render component with props 1`] = `
.c0 {
  color: #ffffff;
  margin: 0;
  font-size: 36px;
  font-weight: bold;
  line-height: 45px;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0:hover {
  color: #4C5667;
}

<a
  class="c0"
  color="contrastText"
  href="/"
  target="_blank"
>
  Home
</a>
`;
